Add howto-checkbox

robdodson opened the pull request:
For fun, here's a checkbox implementation. Just need to drop in comments and tests. I've got a lot of flights coming up, perfect time to write tests! ✈️

Something worth noting, I think this is the first of our elements to expose an attribute and a property. We may need to revisit the others at some point to give them similar APIs.

@surma @notwaldorf PTAL

robdodson commented:
ooh yeah i need to add labeling of some kind... will get to that... 😸

surma commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,77 @@
+    _toggleChecked() {
+      const newState = !this.checked;
+      this.checked = newState;
+      this.dispatchEvent(new CustomEvent('check-changed', {
```
Wouldn’t it be good to make this compatible to a normal `input[type=checkbox]` and fire a `change` event that bubbles?

surma commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,77 @@
+
+      switch (event.keyCode) {
+        case KEYCODE.SPACE:
+          this._toggleChecked();
```
Do you wanna do `[disabled]`, too?

robdodson replied:
should be fixed in current version

robdodson commented:
@surma PTAL

robdodson commented:
Added the `disabled` property

surma commented:
You just... deleted my commits 😭 

I added the commit that adds component-specific assets to the output folder (in this case the SVGs).

robdodson commented:
@alice whenever you have a chance, if you could review the element (https://github.com/GoogleChrome/dash-elements/pull/30/files#diff-681557de3fabc0a0d6f8cb334aabc281) that would be awesooooome

robdodson commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,123 @@
+     * user has predefined an `aria-checked` state or `tabindex`.
+     */
+    connectedCallback() {
+      this.setAttribute('role', 'checkbox');
```
I guess technically I should check to see if the user has applied their own role and respect that

azakus replied:
Guard this on the first `connectedCallback` call

robdodson commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    /**
+     * `attributeChangedCallback` watches for changes to the `checked`
```
todo: add comment mentioning it watches `[disabled]` as well

robdodson replied:
fixed

azakus commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,123 @@
+      return ['checked', 'disabled'];
+    }
+
+    constructor() {
```
No need to have `constructor` if you're just calling `super`.

azakus commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,123 @@
+      if (!this.hasAttribute('tabindex'))
+        this.setAttribute('tabindex', 0);
+
+      this.addEventListener('keydown', this._onKeyDown);
```
The event listeners use `this`, but you didn't bind the function or use arrows. 👃

robdodson replied:
do I need to bind it if the listener is always going to be called in the context of this class instance?

azakus replied:
Nah, I just always do it, but that's not necessary 👍

azakus commented on elements/dash-checkbox/images/checked-checkbox.svg:
```diff
@@ -0,0 +1,22 @@
+    <desc>Created with Sketch.</desc>
+    <defs></defs>
+    <g id="Material-Design-Stickersheet" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
+        <g id="components" transform="translate(-396.000000, -226.000000)" fill="#009688">
```
Too much precision for an 18x18 image. Maybe use svgo?

robdodson commented on elements/dash-tabs/dash-tabs.e2etest.js:
```diff
@@ -102,6 +102,7 @@ describe('dash-tabs', function() {
     async function() {
       const lastTab =
         await this.driver.findElement(By.css('[role=tab]:last-of-type'));
+      // FIXME: This line is missing an await
```
oh yeah @surma this was a note I left to remind  you that I think this tabs test needs to await reading the attributes.

surma replied:
Well... I guess the return value did not equal `true`, indeed 🤔 Thanks!

alice commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,121 @@
+
+    /**
+     * The `checked` property reflects its state to the `aria-checked`
+     * attribute.
```
You're not doing full JSDoc with params and return types?

robdodson replied:
we haven't been. We have a custom parser that extracts the comments for side by side documentation.

alice replied:
Can I see what that looks like?

alice replied:
nm found it. Ok, so no help on parameter types then.

robdodson replied:
Not currently. We could add something like that down the line. cc @surma

surma replied:
We could. Not sure it adds much value since we are not generating API documentation. 

@notwaldorf @robdodson Opinions?

alice commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,121 @@
+     * attribute.
+     */
+    set checked(isChecked) {
+      this.setAttribute('aria-checked', isChecked);
```
Not supporting `aria-checked="mixed"` then? https://www.w3.org/TR/wai-aria-1.1/#aria-checked

robdodson replied:
nope

alice commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,121 @@
+     * attribute. A disabled checkbox will be visible, but no longer operable.
+     */
+    set disabled(isDisabled) {
+      this.setAttribute('aria-disabled', isDisabled);
```
Does this have any bearing on focusability?

robdodson replied:
yeah I guess it should

alice commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -103,6 +103,11 @@
      */
     set disabled(isDisabled) {
       this.setAttribute('aria-disabled', isDisabled);
+      if (isDisabled) {
```
Check or call this method in `connectedCallback()` as well?

robdodson replied:
I'm a little sleepy but I _think_ we don't need to because it'll get triggered by `attributeChangedCallback` when the element boots if the user has added `[disabled]`. If the user has NOT added disabled, then what's in connectedCallback will happen first and set tabindex to 0

alice replied:
So `attributeChangedCallback` gets called for all attributes on boot? Why do you handle any in `connectedCallback` then?

Maybe add a test for this anyway? i.e.
```html
<dash-checkbox aria-disabled="true">  <!-- assert this is not focusable -->
```

surma replied:
+1 for test

robdodson replied:
`attributeChangedCallback` only gets called for attributes listed in `observedAttributes`. It seems it only gets called if something is present so if I didn't put `disabled` on the element it would not be called.

robdodson commented:
I added two new methods to safely modify attributes without triggering attributeChangedCallback https://github.com/GoogleChrome/dash-elements/pull/30/files#diff-681557de3fabc0a0d6f8cb334aabc281R79

I'm wondering if folks would prefer that I always call these methods? For instance, `tabindex` is not an observedAttribute so it will never trigger attributeChangedCallback. But for consistency sake I could still call `_safelySetAttribute` when I modify it. What do you all think? @surma @alice

surma commented on elements/dash-checkbox/dash-checkbox.js:
```diff
@@ -0,0 +1,155 @@
+    _safelySetAttribute(attr, value) {
+      if (this._safelyModifyingAttribute) return;
+      this._safelyModifyingAttribute = true;
+      this.setAttribute(attr, value);
```
Is this guaranteed to work in all CE implementations? I.e. could `attributeChangedCallback` be called in a microtask?

Also, what about removing the attribute from `observedAttributes` before changing the value? Would that work?

surma replied:
I realized you can’t change observedAttributes at runtime.

Looking at the [spec](https://w3c.github.io/webcomponents/spec/custom/#custom-element-reactions), it _seems_ like this implementation is fine. Not entirely sure tho. @azakus WDYT?

> The way in which custom element reactions are invoked is done with special care, to avoid running author code during the middle of delicate operations. Effectively, they are delayed until "just before returning to user script". This means that for most purposes they appear to execute synchronously, but in the case of complicated composite operations (like cloning, or range manipulation), they will instead be delayed until after all the relevant user agent processing steps have completed, and then run together as a batch.

robdodson commented:
k @surma @alice i think this is ready for review again

robdodson commented:
I think I've addressed all outstanding issues.

I added a label to the demo as a near term fix. Maybe a good next element to tackle would be a `<howto-label>`.

@surma PTAL. I'd love to merge this tomorrow if we can.

alice commented on elements/howto-checkbox/howto-checkbox.js:
```diff
@@ -0,0 +1,144 @@
+     * `attributeChangedCallback` watches for changes to the `checked`
+     * and `disabled` attributes and reflects their states to the corresponding
+     * ARIA attributes. It will be called at startup time if either attribute
+     * has been set. Because both `checked` and `disabled` are booleans, the
```
Maybe pull "Because...present" into an inline comment?

alice commented on elements/howto-checkbox/howto-checkbox.js:
```diff
@@ -0,0 +1,144 @@
+     * The `checked` property reflects its state to the `checked`
+     * attribute.
+     */
+    set checked(isChecked) {
```
Maybe move this (and all the other setters and getters) up earlier to make the `_toggleChecked()` -> `set checked()` -> `attributeChangedCallback()` sequence easier to follow?

alice commented on elements/howto-checkbox/howto-checkbox.unittest.js:
```diff
@@ -0,0 +1,31 @@
+      expect(this.checkbox.getAttribute('tabindex')).to.equal('0');
+    });
+
+    it('should toggle `checked` and `aria-checked` when calling ' +
```
Add a similar test for `checkbox.checked = false`/`true`?

alice commented on elements/howto-checkbox/howto-checkbox.unittest.js:
```diff
@@ -0,0 +1,31 @@
+    it('should toggle `checked` and `aria-checked` when calling ' +
+      '`_toggleChecked`', function() {
+        expect(this.checkbox.checked).to.equal(false);
+        this.checkbox._toggleChecked();
```
I assume this method isn't normally public..?

surma commented:
Closing this one in favor of #41 then :)
